pkg/agent/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,451 @@
+		return ""
+	}
+	return service.Status.LoadBalancer.Ingress[0].IP
+}
+
+func (c *ServiceExternalIPController) syncService(key apimachinerytypes.NamespacedName) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Service for %s. (%v)", key, time.Since(startTime))
+	}()
+
+	service, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return c.deleteService(key)
+		}
+		return err
+	}
+
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return c.deleteService(key)
+	}
+
+	state, exist := c.getServiceState(service)
+	currentExternalIP := c.getServiceExternalIP(service)
+	if exist && state.ip != currentExternalIP {
+		if err := c.deleteService(key); err != nil {
+			return err
+		}
+	}
+
+	ipPoool := service.ObjectMeta.Annotations[types.ExternalIPPoolAnnotationKey]
+	if currentExternalIP == "" || ipPoool == "" {
+		return nil
+	}
+
+	selectNode := true
+	var filters []func(string) bool
+	if service.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+		nodes, err := c.nodesHasHealthyServiceEndpoint(service)
+		if err != nil {
+			return err
+		}
+		// Avoid unnecessary migration caused by Endpoint changes.
+		if exist && c.cluster.AliveNodes().Has(state.assignedNode) && nodes.Has(state.assignedNode) {
+			selectNode = false
+		} else {
+			filters = append(filters, func(s string) bool {
+				return nodes.Has(s)
+			})
+		}
+	}
+
+	if selectNode {
+		nodeName, err := c.cluster.SelectNodeForIP(currentExternalIP, ipPoool, filters...)
+		if err != nil {
+			return err
+		}
+		klog.InfoS("SelectNodeForIP", "nodeName", nodeName, "currentExternalIP", currentExternalIP, "ipPoool", ipPoool)
+		state = externalIPState{
+			ip:           currentExternalIP,
+			assignedNode: nodeName,
+		}
+		c.saveServiceState(service, state)
+	}
+
+	serviceToUpdate := service.DeepCopy()
+	serviceToUpdate.Status.LoadBalancer.Ingress[0].Hostname = state.assignedNode
+
+	if state.assignedNode == c.nodeName {
+		if _, err = c.client.CoreV1().Services(serviceToUpdate.Namespace).UpdateStatus(context.TODO(), serviceToUpdate, v1.UpdateOptions{}); err != nil {
+			return err
+		}
+		return c.ipAssigner.AssignIP(currentExternalIP)
+	}
+	return c.ipAssigner.UnassignIP(currentExternalIP)
+}
+
+// nodesHasHealthyServiceEndpoint returns the set of Nodes which has at least one healthy endpoint.
+func (c *ServiceExternalIPController) nodesHasHealthyServiceEndpoint(service *corev1.Service) (sets.String, error) {
+	nodes := sets.NewString()
+	endpoints, err := c.endpointsLister.Endpoints(service.Namespace).Get(service.Name)
+	if err != nil {
+		return nodes, err
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, ep := range subset.Addresses {
+			if ep.NodeName == nil {
+				continue
+			}
+			nodes.Insert(*ep.NodeName)
+		}
+		for _, ep := range subset.NotReadyAddresses {
+			if ep.NodeName == nil {
+				continue
+			}
+			nodes.Delete(*ep.NodeName)
```

Comment:
Why exclude a node when it includes any not ready address? There could be ready addresses.